packages/plugins/scripts-runner/src/index.ts

```diff
@@ -0,0 +1,208 @@
+import { eachSeries } from 'async';
+import { promises as fsPromises, stat, Stats } from 'fs';
+import path from 'path';
+import { Embark, Callback, Plugins } from 'embark-core';
+import { Logger } from 'embark-logger';
+import { dappPath } from 'embark-utils';
+import Web3 from "web3";
+
+export enum ScriptsRunnerCommand {
+  Execute = 'scripts-runner:execute'
+}
+
+export enum ScriptsRunnerEvent {
+  Executed = 'scripts-runner:script:executed'
+}
+
+export interface ScriptDependencies {
+  web3: Web3 | null;
+  contracts: any;
+  logger: Logger;
+}
+
+export interface ExecuteOptions {
+  target: string;
+  dependencies: ScriptDependencies;
+  forceTracking: boolean;
+}
+
+enum FileType {
+  SymbolicLink = 'symbolic link',
+  Socket = 'socket',
+  Unknown = 'unknown'
+}
+
+class UnsupportedTargetError extends Error {
+
+  name = 'UnsupportedTargetError';
+
+  constructor(public stats: Stats) {
+    super();
+    // We can't access `this` before `super()` is called so we have to
+    // set `this.message` after that to get a dedicated error message.
+    this.setMessage();
+  }
+
+  private setMessage() {
+    let ftype = FileType.Unknown;
+    if (this.stats.isSymbolicLink()) {
+      ftype = FileType.SymbolicLink;
+    } else if (this.stats.isSocket()) {
+      ftype = FileType.Socket;
+    }
+    this.message = `Script execution target not supported. Expected file path or directory, got ${ftype} type.`;
+  }
+}
+
+class ScriptExecutionError extends Error {
+
+  name = 'ScriptExecutionError';
+
+  constructor(public target: string, public innerError: Error) {
+    super();
+    this.message = `Script '${path.basename(target)}' failed to execute. Original error: ${innerError}`;
+  }
+}
+
+export default class ScriptsRunnerPlugin {
+
+  private _web3: Web3 | null = null;
+
+  constructor(private embark: Embark, private plugins: Plugins) {
+    // TODO: it'd be wonderful if Embark called `registerCommandHandlers()` for us
+    this.registerCommandHandlers();
+  }
+
+  registerCommandHandlers() {
+    this.embark.events.setCommandHandler(ScriptsRunnerCommand.Execute, this.execute.bind(this));
+  }
+
+  private get web3() {
+    return (async () => {
+      if (!this._web3) {
+        const provider = await this.embark.events.request2('blockchain:client:provider', 'ethereum');
+        this._web3 = new Web3(provider);
+      }
+      return this._web3;
+    })();
+  }
+
+  private async execute(target: string, forceTracking = false, callback: Callback<any>) {
+    const targetPath = !path.isAbsolute(target) ? dappPath(target) : target;
+    try {
+      const fstat = await fsPromises.stat(targetPath);
+      if (fstat.isDirectory()) {
+        const dependencies = await this.getScriptDependencies();
+        await this.executeAll({ target: targetPath, dependencies, forceTracking },callback);
```

Comment:
Why does `executeAll` fire a callback and return a promise? Same question re: `executeSingle`.

Related to that, will the callback fire multiple times in the context of `executeAll`?

Reply:
> Why does executeAll fire a callback and return a promise? Same question re: executeSingle.

That's a left over. As mentioned in https://github.com/embarklabs/embark/pull/2253/files#r374630373 both, `executeSingle` and `executeAll` could be exposed as individual commands (right not it's just `execute` which then determines when all or single have to run). If both individual APIs are exposed as commands, they have to receive a callback (https://www.pivotaltracker.com/story/show/171000554).

So yes, the `await` here isn't necessary, we can just rely on `callback`. Will update.

> Related to that, will the callback fire multiple times in the context of executeAll?

It won't. `executeAll()` passes an `eachSeries()` callback to `executeSingle`. The `callback` passed to `executeAll` is only called once when the series is done: https://github.com/embarklabs/embark/pull/2253/files#diff-2c89f5499b488179b7c77219d3ec509cR173